packages/jaeger-ui/src/model/link-patterns.js

```diff
@@ -0,0 +1,183 @@
+        if (i % 2 === 0) {
+          text += templateSplit[i];
+        } else {
+          text += encodeFn(args[templateSplit[i]]);
+        }
+      }
+      return text;
+    },
+  };
+}
+
+export function createTestFunction(entry) {
+  if (typeof entry === 'string') {
+    return arg => arg === entry;
+  }
+  if (Array.isArray(entry)) {
+    return arg => entry.indexOf(arg) > -1;
+  }
+  if (entry instanceof RegExp) {
+    return arg => entry.test(arg);
+  }
+  if (typeof entry === 'function') {
+    return entry;
+  }
+  if (entry == null) {
+    return () => true;
+  }
+  throw new Error(`Invalid value: ${entry}`);
+}
+
+const identity = a => a;
+
+export function processLinkPattern(pattern) {
+  try {
+    const url = processTemplate(pattern.url, encodeURIComponent);
+    const text = processTemplate(pattern.text, identity);
+    return {
+      object: pattern,
+      type: createTestFunction(pattern.type),
+      key: createTestFunction(pattern.key),
+      value: createTestFunction(pattern.value),
+      url,
+      text,
+      parameters: _uniq(url.parameters.concat(text.parameters)),
+    };
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error(`Ignoring invalid link pattern: ${error}`, pattern);
+    return null;
+  }
+}
+
+export function getParameterInArray(name, array) {
+  if (array) {
+    return array.find(entry => entry.key === name);
+  }
+  return undefined;
+}
+
+export function getParameterInAncestor(name, spans, startSpanIndex) {
+  let currentSpan = { depth: spans[startSpanIndex].depth + 1 };
+  for (let spanIndex = startSpanIndex; spanIndex >= 0; spanIndex--) {
+    const nextSpan = spans[spanIndex];
+    if (nextSpan.depth < currentSpan.depth) {
+      currentSpan = nextSpan;
+      const result =
+        getParameterInArray(name, currentSpan.tags) || getParameterInArray(name, currentSpan.process.tags);
+      if (result) {
+        return result;
+      }
+    }
+  }
+  return undefined;
+}
+
+export function callTemplate(template, data) {
+  return template.template(...template.parameters.map(param => data[param]));
+}
+
+export function computeLinks(linkPatterns, trace, spanIndex, items, itemIndex) {
+  const item = items[itemIndex];
+  const span = trace.spans[spanIndex];
+  let type = 'logs';
+  const processTags = span.process.tags === items;
+  if (processTags) {
+    type = 'process';
+  }
+  const spanTags = span.tags === items;
+  if (spanTags) {
+    type = 'tags';
+  }
+  const result = [];
+  linkPatterns.forEach(pattern => {
+    if (pattern.type(type) && pattern.key(item.key, item.value, type) && pattern.value(item.value)) {
+      let parameterValues = {};
+      pattern.parameters.every(parameter => {
+        let entry = getParameterInArray(parameter, items);
+        if (!entry && !processTags) {
+          // do not look in ancestors for process tags because the same object may appear in different places in the hierarchy
```

Comment:
Given this comment, why are process tags always searched when looking through ancestors? Seems like the only process tags that are skipped are those on the current span.

Also, why is it relevant that process tags can be repeated? A process will match the first time it's encountered or it will fail every time it's tested?

Reply:
The `getLinks` function caches results in the `alreadyComputed` weak map.
The key used in that map is the object (with the `key` and `value` properties) that will be displayed as a link (if a link has to be displayed).
If that object is reused at different places in the hierarchy (which is the case for process tags, as the same process object can be reused for multiple spans), a call to `getLinks` with an object that is already in the cache could get the result of a previous call to `getLinks` from a different place in the hierarchy. If a process tag could use values from an ancestor, the ancestor from which the value could be found and the corresponding value could be different between the two places in the hierarchy, which would make `getLinks` return an invalid value.
Another solution (instead of preventing process tags links from using values from ancestors) would be to have a cache that also stores the current span (so we would need a cache with 2 keys), but that would be more complex and I am not sure if it is actually needed to refer to ancestor values from a process tag.

Reply:
I will explain in Javascript:

```js
// Let's say there is one tag like this:

let myProcessTag = {
  key: 'myProcessInfo',
  value: 'myProcessValue'
};

// The tag is used in a process:

let myProcess = {
  tags: [myProcessTag]
};

// The same process reference is used in multiple spans in the trace hierarchy:
let trace = {
  spans: [{
    depth: 0
  }, {
    depth: 1,
    tags: [{
      key: 'myParentKey',
      value: 'parentValue1'
    }]
  }, {
    depth: 2,
    process: myProcess
  }, {
    depth: 1,
    tags: [{
      key: 'myParentKey',
      value: 'parentValue2'
    }]
  }, {
    depth: 2,
    process: myProcess
  }]
};

// So myProcess is used in trace.spans[2], which is the child of trace.spans[1]
// and myProcess is also used in trace.spans[4], which is the child of trace.spans[3]

// Now, let's suppose that it is allowed for a link of a process tag to refer to tags from ancestors
// and someone defined the configuration like this:
JAEGER_CONFIG = {
  "linkPatterns": [{
    "key": "myProcessInfo",
    "url": "http://example.org/?myProcessValue=#{myProcessInfo}&myParentValue=#{myParentKey}",
    "text": "Where will this link go?"
  }]
};

// This link on the process tag refers to a tag from an ancestor
// (which is the thing I am not allowing in the current code).

// Then, if we call computeLinks with trace.spans[2] and myProcessTag, we should get:
[{
  "url": "http://example.org/?myProcessValue=myProcessValue&myParentValue=parentValue1",
  "text": "Where will this link go?"
}]

// Now, if we call computeLinks with trace.spans[4] and the same myProcessTag, we should get:
[{
  "url": "http://example.org/?myProcessValue=myProcessValue&myParentValue=parentValue2",
  "text": "Where will this link go?"
}]

// These two values are different because the same tag is at two different places in the spans hierarchy,
// and the value for myParentKey is taken
// - in the first case from trace.spans[1].tags[0]
// - in the second case from trace.spans[3].tags[0]

// Now the getLink function returned by createGetLinks uses the tag object as a key in the cache.
// In our case, the tag object is myProcessTag.

// So if we first call the getLink function returned by createGetLinks for trace.spans[2] and myProcessTag :
result = cache.get(myProcessTag); // result is undefined
result = computeLinks(...);
// which means:
result = [{
  "url": "http://example.org/?myProcessValue=myProcessValue&myParentValue=parentValue1",
  "text": "Where will this link go?"
}];
cache.set(myProcessTag, result); // result is cached

// now if we call the getLink function for trace.spans[4] and myProcessTag :
result = cache.get(myProcessTag); // result is already computed
return [{
  "url": "http://example.org/?myProcessValue=myProcessValue&myParentValue=parentValue1",
  "text": "Where will this link go?"
}]; // now the return value is wrong for trace.spans[4]
```

I hope this explanation is clearer!
I do not have the requirement of having links on process tags that use values from ancestor spans, so I chose the simplest solution of not allowing it. Do you think it would be better (for consistency or some use cases that I don't know) to allow this and then to use as the key in the cache both the span and the tag object (for example, with a cache based on nested WeakMaps)?

Reply:
Great explanation, I think I've got it... Don't look in ancestors for process tags because the same process object, and therefore tags, can be used for many spans. In this case, searching a span's ancestors could yield different results (if they weren't cached). But, as the cache is keyed on the initial, shared, process tag, all spans with that process would end up getting the same results even if they have different ancestors?

Reply:
Exactly!